### ISSUE TITLE
No unification of build and intellisense diagnostics in VS2017 (was in VS2015)

### ISSUE THREAD

**Issue by dsyme** [Microsoft/visualfsharp#3047]
There are likely [multiple issues with the Visual Studio Error list](https://github.com/Microsoft/visualfsharp/issues/2550)..  I'm splitting out one known repro for a basic one, which is the lack of "Error Unification" between build and intellisense errors.  

1. open any solution, e,g,

    devenv tests\projects\misc\TestProjectChanges\TestProjectChanges.sln

2,. add an error, watch intelliense error appear

3. do a build, wathc duplicate copy of "build" error appear

4. fix the error - the "intellisense" copy goes away, but the red suiglly stays

5. do a build - the "build" error goes away

Although logical, it is painful to have the red squiggly stay around even after you've fixed the error.  In practice this means that once you do a build, and there are errors, you have to keep doing builds do clear all your red-squigglies, or else resort to showing Intellisense errors only.  

This is a regression (or at least a change) from how F# worked in VS2015 and before, when build and intellisense errors were "unified" and fixing an error made it disappear from the error list - the user just was not particularly aware of the distinction between build and intellisense errors.

Now I'm not sure how deeply this distinction is built into Roslyn - it may not be something we control anymore - advice welcome please @Pilchie  @DustinCampbell 

Note that the diagnostic ID is also missing for the build error - this could be the underlying problem since some comments in the F# language service code imply that error de-duplication should be ocurring.  It's possible this is a problem with whichever error parser is being used by default 

![image](https://cloud.githubusercontent.com/assets/7204669/26025097/0b834f10-37d8-11e7-80ee-a4e4322aa41b.png)

Here is the output from the build, it contains ``FS0039``

```
3>C:\GitHub\dsyme\visualfsharp\tests\projects\misc\TestProjectChanges\TestProjectChanges\Program.fs(7,58): error FS0039: The field, constructor or member 'PropertyA' is not defined. Maybe you want one of the following: PropertyAlwaysAvailable
3>Done building project "TestProjectChanges.fsproj" -- FAILED.
```

**Comment by dsyme** [comment]
(Note, I haven't yet compared to C# roslyn behavior)

**Comment by BentTranberg** [comment]
@dsyme, I compared C# and F# in VS 2015 and VS 2017, and the results is in a comment a bit down in #2548. I don't understand that much of this, but it looks to me there's pretty much the same issue in VS 2015, but just not that obvious to the user.

**Comment by dsyme** [comment]
@BentTranberg Ah yes, thanks.  Good analysis.  TBH I much prefer the "unified error list" model of F# in VS015 (all errors are just one kind of error), I wonder why C#/roslyn didn't adopt this model.

**Comment by dsyme** [comment]
One of the odd things here is that [this recently added code ](https://github.com/Microsoft/visualfsharp/blame/ce473ddab307c1bbbf027207badf086e71a238ff/vsintegration/src/FSharp.Editor/Common/RoslynHelpers.fs#L108) seems to indicate that we are trying to do error unification.  

        // Normalize the error message into the same format that we will receive it from the compiler.
        // This ensures that IntelliSense and Compiler errors in the 'Error List' are de-duplicated.
        // (i.e the same error does not appear twice, where the only difference is the line endings.)

However the displayed error message in the error list above still has newline characters.  Odd, will debug through this

**Comment by saul** [comment]
@dsyme we are trying but we aren't reporting build errors through Roslyn, so it doesn't see the build/intellisense errors as identical (as the former don't have error IDs amongst other data). I believe I put more info in my PR: https://github.com/Microsoft/visualfsharp/pull/2839#issuecomment-294573571

**Comment by dsyme** [comment]
> @dsyme we are trying but we aren't reporting build errors through Roslyn, so it doesn't see the build/intellisense errors as identical (as the former don't have error IDs amongst other data). I believe I put more info in my PR: #2839 (comment)

Thanks, that's very helpful!  Here's a copy of the relevant comment:

> @heejaechang I can answer that. F# build errors are passed to the IDE via the ErrorListProvider, which is wrapped by a Microsoft.VisualStudio.Shell.TaskProvider. See DocumentTask.cs@506 in the F# code.
> 
> I think the correct solution for this is to move over to Roslyn's ProjectExternalErrorReporter. We currently don't report any errors through this, but this class implements IVsLanguageServiceBuildErrorReporter2 so we can report build errors through the ReportError2 method. This is what the Roslyn project system does.

**Comment by dsyme** [comment]
More notes. The current F# project system has ``IDEBuildLogger`` which is added to MSBuild loggers in ``DoMSBuildSubmission``

                    loggerList.Add(buildLogger);

MSBuild then raises events and these get pushed to ``ErrorHandler`` and ``WarningHandler`` and thus to ``AddToErrorList``.   This creates a ``DefaultCompilerError`` (which does contain the error code) and a ``DocumentTask`` which doesn't. The task gets registered using ``taskReporter.AddTask``.  As @saul mentions, this last call should be removed and we should probably use ``ProjectExternalErrorReporter``

**Comment by dsyme** [comment]
@Pilchie I'd appreciate some guidance here: does Roslyn come with any mechanism for listening to build output (e.g. from an MSBuild build logger) and creating matching tasks from the Error and Warning items that come from MSBuild?  And matching those to the text being shown in the output window? 

The context is that F# intellisense errors (via roslyn) and the F# build errors (via non-roslyn) are not being de-duplicated. This is because of subtle differences in the content of the error/task items being produced by these two paths (slightly different text in multi-line error messages, and a missing ID for the non-Roslyn errors).  Ideally we want to produce all errors via roslyn to better ensure consistency.  But we don't want to reinvent the wheel here.

**Comment by dsyme** [comment]
@saul @Pilchie  I notice we do already create our own [ProjectExternalErrorReporter](https://github.com/Microsoft/visualfsharp/blob/43c90f8554ec534c35bdea3b730c82f31b46d0a1/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L356), perhaps we just have to use that

Glancing around the Roslyn code I see [GetExternalErrorReporter](https://github.com/dotnet/roslyn/blob/38afa69825db300e58aa6416dbb2c14ab340da7f/src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs#L99) but also note there is no F# mapping [here](https://github.com/dotnet/roslyn/blob/38afa69825db300e58aa6416dbb2c14ab340da7f/src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs#L26), which must be why we create our own. It would be helpful to start removing the C#/VB-specific things from Roslyn and just make it a multi-language framework.

**Comment by dsyme** [comment]
@saul @Pilchie 

I took a look at what it would take to junk the existing F# code for build task creation and use ProjectExternalErrorReporter.  A draft of what would be needed is here: 
 https://github.com/Microsoft/visualfsharp/compare/master...dsyme:eld?expand=1

@saul If you get the chance could you take a glance over that draft and see what you think?  The idea is to literally delete the use of TasKReporter and TaskProvider in the ProjectSystem.Base and replace it with a propagated ``IVsLanguageServiceBuildErrorReporter2`` (i.e. a ProjectExternalErrorReporter)

**Comment by Pilchie** [comment]
@heejaechang has pretty much done everything related to the Error List since Roslyn started, and I'll admit that I've lost track of it.

HeeJae - can you confirm my understanding is correct?

IIRC, Roslyn has no ability to insert msbuild loggers or anything, but the C# project system looks for error output and creates the "Build" tasks.  Roslyn only reports the "IntelliSense" errors from it's in-memory model.

**Comment by dsyme** [comment]
@saul @heejaechang @Pilchie  I updated the branch above, https://github.com/Microsoft/visualfsharp/compare/master...dsyme:eld?expand=1.  

However I get a crash when using it.  I believe the ``CompilerDiagnosticAnalyzer`` service has not been implemented for F#.  We do implement ``DocumentDiagnosticAnalyzer``.  I'll try to implement ``CompilerDiagnosticAnalyzer``  tomorrow, I presume it is not too hard.

Here are the details how I got to this conclusion if you're interested... Creating the error manager makes us fall active paths to this FailFast check https://github.com/dotnet/roslyn/blob/175c1605927859bc3cfcd4a84b720b27c4054ba9/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.cs#L139. This is because this code:

                    var compilerAnalyzer = _analyzerManager.GetCompilerDiagnosticAnalyzer(project.Language);

is returning null for F#. This is because:

            return _hostAnalyzerManager.GetCompilerDiagnosticAnalyzer(language);

is returning null. The _hostAnalyzerManager comes from [this](https://github.com/dotnet/roslyn/blob/664f2634a4a0ea25d6243becd46147333b7f38a0/src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService.cs#L31).  So [this](https://github.com/dotnet/roslyn/blob/1ca38b7717b4079c9d1069552d801bb59d212542/src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs#L252) is returning null, because [_compilerDiagnosticAnalyzerMap.TryGetValue](https://github.com/dotnet/roslyn/blob/1ca38b7717b4079c9d1069552d801bb59d212542/src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs#L255) is returning null,

**Comment by dsyme** [comment]
Ouch, ComppilerDiagnosticAnalyzer is internal in Microsoft.CodeAnalysis.dll, and FSharp.Editor doesn't have an IVT attribute to that DLL (it has [IVT attributes](https://github.com/dotnet/roslyn/search?utf8=%E2%9C%93&q=FSharp.Editor&type=) to many other DLLs).

@Pilchie @cartermp  This is a bit of a show-stopper on fixing this regression from VS2015

**Comment by Pilchie** [comment]
@dsyme - Roslyn's ErrorList implementation *also* doesn't have IVT to Microsoft.CodeAnalysis.dll.  That dll is part of our compiler layer, and none of the VS assemblies have IVT to it.

**Comment by Pilchie** [comment]
I expect we actually depend on the [`CompilerDiagnosticAnalyzer`](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs,56) from `Microsoft.CodeAnalysis.EditorFeatures.dll`, and F# should have IVT to that.

**Comment by dsyme** [comment]
@Pilchie Thanks.  I think it has an IVT (though I don't seem to be able to see [DefaultDiagnosticAnalyzerService](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DefaultDiagnosticAnalyzerService.cs,18) which is strange).  Though even if I could get one of those I'm not yet sure how would I pass it in and associate it with the F# language name - perhaps it's easy

**Comment by dsyme** [comment]
@Pilchie Poking around I also notice scary not-friendly-to-F# things like [this](https://github.com/dotnet/roslyn/blob/0d6f96aa4bd85f4c616c616e0a79e41adf2283c2/src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs#L92) (found that via a code search) which would indicate it's not going to recognize anything as a CompilerDiagosticAnalyzer easily.  That's used [here](https://github.com/dotnet/roslyn/blob/1ca38b7717b4079c9d1069552d801bb59d212542/src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs#L397)

It's late and I'm kind of lost in this stuff.  If you can see a way to solve the basic problem of reusing Roslyn's implementation of ProjectExternalErrorReporter (which seems to require a CompilerDiagosticAnalyzer) then that would be very helpful.

**Comment by dsyme** [comment]
@saul To recap - I think [my prototype branch to use ReportError2 via our ProjectExternalErrorReporter](https://github.com/Microsoft/visualfsharp/compare/master...dsyme:eld?expand=1) is quite close to working but we've hit this internal Roslyn problem while trying to use the Roslyn ``ProjectExternalErrorReporter``.    

Perhaps there is some other path we can take like our own implementation of the VS interface for ReportError2 though there's a significant amount of code involved in that.

I'm a bit concerned that there are a few landmines for F# scattered over the Roslyn codebase (e.g. places where if-C#-if-VB is hardwaired).  If we stick to the well-trodden path that TypeScript has already used we should be OK I suppose, but it's not easy to tell what that path is.

**Comment by dsyme** [comment]
@saul @Pilchie I've added a Roslyn issue here which I think summarizes things: https://github.com/dotnet/roslyn/issues/19547

**Comment by heejaechang** [comment]
@dsyme can you try this? https://github.com/dotnet/roslyn/pull/19561 to see whether this fix your issue?

**Comment by dsyme** [comment]
@heejaechang Thanks!  I'll aim to give that a go tomorrow UK time, if not the next day

**Comment by dsyme** [comment]
An update on this: we've made good progress towards unifying the build and intellisense error lists , see discussion on https://github.com/dotnet/roslyn/pull/19561.  This even appears to work for multi-line error messages which I had expected to be problematic

This requires a Roslyn fix, so we have to see when that makes it in - hopefully soon!  We will also **require** the Roslyn fix to be in place, or Visual Studio will crash when it hits the call to FailFast.

**Comment by saul** [comment]
@dsyme Glad to see we're making progress on this. Also the multi-line messages are already identical, I made sure of that in my PR when I attempted to fix it 👍

**Comment by dsyme** [comment]
> Also the multi-line messages are already identical, I made sure of that in my PR when I attempted to fix it 

Nah :) The ones coming from the build do **not** have line breaks.  See the two screen shots below, the first showing the Intellisense error, the second the Build error

![image](https://cloud.githubusercontent.com/assets/7204669/26167762/a7b0432a-3b2f-11e7-8738-d80b0dc63a6b.png)

![image](https://cloud.githubusercontent.com/assets/7204669/26167779/b17f32ee-3b2f-11e7-9467-51123e91d018.png)

**Comment by saul** [comment]
Hmm then something isn't right... or the build errors are being reported to the IDE in a different way now? There is code that explicitly replaces the fake newline character with the real one before passing it to the platform: https://github.com/Microsoft/visualfsharp/blob/ce473ddab307c1bbbf027207badf086e71a238ff/vsintegration/src/FSharp.Editor/Common/RoslynHelpers.fs#L108-L111

**Comment by dsyme** [comment]
@saul That's right for Intellisense errors.  The question is build errors, which go via the IDEBuildLogger class. 

The screenshots above are for [my branch](https://github.com/Microsoft/visualfsharp/compare/master...dsyme:eld?expand=1
) when using [a fixed Roslyn](https://github.com/dotnet/roslyn/pull/19561) 

When using master, you get this:

![image](https://cloud.githubusercontent.com/assets/7204669/26168434/f35aeb0c-3b31-11e7-84ef-19211064bef8.png)

**Comment by saul** [comment]
In which case, the errors should definitely go through `ErrorLogger.NormalizeErrorString |> ErrorLogger.NewlineifyErrorString`. 

Your last screenshot also doesn't agree with the screenshot I put on the PR, which does have new lines in both cases. Maybe it's something specific to the suggested names? Either way it should have new lines!

![image](https://cloud.githubusercontent.com/assets/64654/26168667/da357bbe-3b32-11e7-8bba-d064ac35a7bc.png)

**Comment by dsyme** [comment]
> In which case, the errors should definitely go through ErrorLogger.NormalizeErrorString |> ErrorLogger.NewlineifyErrorString.

Yes, I was wondering that.  Though don't forget the error text is scraped from the output of fsc.exe  by MSBuild (with --flaterrors).  Does that even have the magic characters needed?

**Comment by dsyme** [comment]
@brettfo The fix is now in roslyn - https://github.com/dotnet/roslyn/pull/19561. Could you update this issue when we have those Roslyn binaries in this repo please?   THanks!

**Comment by Pilchie** [comment]
@dsyme We don't install roslyn binaries here.  As there are no API changes involved in the Roslyn change, this should just start working on newer builds of VS with that Roslyn change.

**Comment by dsyme** [comment]
@Pilchie Ah. The problem is that we can't use the functionality until we can be sure we have the fix - if we ended up running against the old binaries we bring down the whole VS process because we end up hitting the FailFast here (I suggested a more gracious error handling technique would have been sensible :) ) 

https://github.com/dotnet/roslyn/blob/175c1605927859bc3cfcd4a84b720b27c4054ba9/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.cs#L139

**Comment by Pilchie** [comment]
In that case I would recommend either:

1. Holding the F# change until there is a public build with the fix, which is unforunate.
2. Conditioning the F# change on the file version of the Roslyn binary being higher than the first Roslyn signed build with this fix (looks like that's 20170524.5) and filing an issue to remove the lightup check once F# depends on a public version with the fix.

**Comment by dsyme** [comment]
> 2.	Conditioning the F# change on the file version of the Roslyn binary being higher than the first Roslyn signed build with this fix (looks like that's 20170524.5) and filing an issue to remove the lightup check once F# depends on a public version with the fix.

That sounds plausible

**Comment by dsyme** [comment]
Fixed by #3071
